src/protocol.rs

```diff
@@ -1,42 +1,541 @@
 use std::net::{Ipv4Addr, Ipv6Addr};
 use std::str::FromStr;
 use std::convert::From;
+use std::fmt;
+use std::hash;
+use std::io;
 use std::io::Cursor;
+use std::io::Read;
+
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
-use cid::Cid;
-use integer_encoding::VarIntWriter;
+use cid;
+use integer_encoding::{VarIntReader, VarIntWriter};
 
 use {Result, Error};
 
+
 ///! # Protocol
 ///!
 ///! A type to describe the possible protocol used in a
 ///! Multiaddr.
 
-macro_rules! build_protocol_enum {
-    {$( $val:expr => $var:ident: $alph:expr, $size:expr, )*} => {
+
+/// Single multiaddress segment with its attached data
+pub trait AddressSegment : fmt::Display + ToString {
```

Comment:
I just thought it would be nice to support rendering multiaddr segements as part of formatting strings. The `ToString` is mostly there to ensure that fact that everything implementing `AddressSegement` also implements `ToString` is highlighted in the documentation and is part of the intended API (instead of being just a fortunate accident).